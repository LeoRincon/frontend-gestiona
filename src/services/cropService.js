import { API_URL, CROPS_PATH } from "../utils/const";

export async function fetchCrops() {
  const sessionUser = sessionStorage.getItem("user_data");

  if (!sessionUser) {
    console.log("Invalid project ID"); // volver al user a la vista de projects
  }

  const { projectId } = JSON.parse(sessionUser);
  const url = `${API_URL}${CROPS_PATH}/project/${projectId}`;

  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const { crops } = await response.json();

  const newData = crops.reduce((acumulador, crop, index) => {
    const { nombre, tipo_siembra, fecha_inicio, area_terreno, unit } = crop;

    const newCrop = {
      id: index + 1,
      nombre,
      tipo_siembra,
      fecha_inicio,
      area_terreno,
      unidad: unit.unidad,
      descripcion: unit.nombre,
    };
    acumulador.push(newCrop);
    return acumulador;
  }, []);

  return newData;
}

export async function createCrop(data) {
  const sessionUser = sessionStorage.getItem("user_data");

  if (!sessionUser) throw new Error("Invalid session");

  const { projectId } = JSON.parse(sessionUser);

  if (!projectId) throw new Error("Invalid Project ID");
  if (!data || typeof data !== "object") throw new Error("Data is required");

  data["proyecto_id"] = projectId;
  data["area_terreno"] = Number(data.area_terreno);

  const url = API_URL + CROPS_PATH;

  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) throw new Error("Error creating Crop at service");

  const { crop } = await response.json();
  const { nombre, tipo_siembra, fecha_inicio, area_terreno, unit } = crop;

  return {
    nombre,
    tipo_siembra,
    fecha_inicio,
    area_terreno,
    unidad: unit.unidad,
    descripcion: unit.nombre,
  };
}

export async function deleteCropByName(nameCrop) {
  if (!nameCrop) throw new Error("ID Crop is required");
  if (typeof nameCrop !== "string") throw new Error("ID Crop must be a string");

  const queryUrl = API_URL + "/query";

  const query = {
    key: "nombre",
    value: nameCrop,
    returnValue: "id",
    table: "cultivo",
  };

  const queryRes = await fetch(queryUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(query),
  });

  if (!queryRes.ok) throw new Error("Error Fetching Id");

  const queryData = await queryRes.json();

  const idCrop = queryData.data.id;

  const url = `${API_URL}${CROPS_PATH}/${idCrop}`;

  const response = await fetch(url, {
    method: "DELETE",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new Error("Error Deleting crop");

  return {
    success: true,
  };
}
