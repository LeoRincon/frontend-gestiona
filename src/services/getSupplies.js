export async function fetchSupplies() {
  const url = "http://localhost:3000/api/v1/supplies"

  const response = await fetch(url,{
    method: "GET",
    headers: {
      "Content-Type": "application/json"
    }
  })

  const {data} = await response.json()
  const newData =data.map((supply) => {
    return {
      id: supply.id,
      nombre: supply.nombre,
      cantidad_disponible: supply.cantidad_disponible,
      fecha_ingreso: supply.fecha_ingreso,
      precio: supply.precio,
      id_inventario: supply.id_inventario,
      id_categoria: supply.id_categoria,
      id_unidad_medida: supply.id_unidad_medida,
    }
  })
  return newData
}

// "id": "b6789012-8def-1234-5678-abc456789abc",
//       "nombre": "Semillas de Ma√≠z",
//       "cantidad_disponible": 200,
//       "fecha_ingreso": "2024-01-01T05:00:00.000Z",
//       "precio": 50,
//       "id_inventario": "a5678901-8abc-1234-5678-abc456789abc",
//       "id_categoria": "e3456789-8abc-1234-5678-abc456789def",
//       "id_unidad_medida": "e2345678-9def-1234-5678-abc456789012"
//     }