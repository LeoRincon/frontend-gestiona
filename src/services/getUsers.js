export async function fetchUsers() {
  const url = "http://localhost:3000/api/v1/users"
  
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json"
    }
  })

  const { data } = await response.json()

  const newdata = data.map((user) => {
    return {
      name: user.nombre,
      email: user.email,
    }
  })
  return newdata
}