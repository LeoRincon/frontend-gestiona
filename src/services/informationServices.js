export async function getInformation() {
  const url = "http://localhost:3000/api/v1/expenses";
  const urlSales = "http://localhost:3000/api/v1/sales";
  const urlProduction = "http://localhost:3000/api/v1/product";
  const urlSupplies = "http://localhost:3000/api/v1/supplies";
  const urlActivitiesManagement = "http://localhost:3000/api/v1/activity_management";

  const response = await fetch(url, { method: "GET", headers: { "Content-Type": "application/json" } });
  const data = await response.json();

  const responseVentas = await fetch(urlSales, { method: "GET", headers: { "Content-Type": "application/json" } });
  const salesData = await responseVentas.json();

  const responseProduccion = await fetch(urlProduction, { method: "GET", headers: { "Content-Type": "application/json" } });
  const productionData  = await responseProduccion.json();

  const responseInsumos = await fetch(urlSupplies, { method: "GET", headers: { "Content-Type": "application/json" } });
  const suppliesData = await responseInsumos.json();

  const responseGestionActividades = await fetch(urlActivitiesManagement, { method: "GET", headers: { "Content-Type": "application/json" } });
  const activityManagementData = await responseGestionActividades.json();

  return {
      expenses: data,
      sales: salesData,
      production: productionData,
      supplies: suppliesData,
      activityManagements: activityManagementData,
  };
}