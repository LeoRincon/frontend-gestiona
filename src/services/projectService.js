import { API_URL, PROJECTS_PATH } from "../utils/const";

export async function fetchProjects() {
  const url = API_URL + PROJECTS_PATH;

  const response = await fetch(url, {
    method: "GET",
    headers:{
      "Content-Type": "application/json"
    }
  })

  const data = await response.json()
  
  return data
}

export async function createProject(project) {
  const url = API_URL + PROJECTS_PATH;

  if (!project || typeof project !== "object")
    throw new Error("The project must be an object");
  
  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(project),
  });

  const data = await response.json();
  return data;
}
