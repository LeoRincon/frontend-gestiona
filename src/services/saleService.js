import { API_URL, SALES_PATH, SEASONS_PATH, UNITS_PATH } from "../utils/const";

const salesUrl = API_URL + SALES_PATH;
const seasonsUrl = API_URL + SEASONS_PATH;
const unitsUrl = API_URL + UNITS_PATH;
const urls = [salesUrl, seasonsUrl, unitsUrl];

export async function createSale(sale) {
  try {
    const response = await fetch(salesUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(sale),
    });
    if (!response.ok) throw new Error("Error creating sale");
    const data = await response.json();

    if (!data.success) throw new Error("Error creating sale");

    return data.sale;
  } catch (error) {
    console.error("Error creating sale: ", error);
  }
}

export async function getSalesData() {
  try {
    const [salesRes, seasonsRes, unitsRes] = await Promise.all(
      urls.map((url) =>
        fetch(url, {
          method: "GET",
          headers: { "Content-Type": "application/json" },
        })
      )
    );
    if (![salesRes, seasonsRes, unitsRes].every((res) => res.ok)) {
      throw new Error("Una o mas repuestas de la API no fueron exitosas");
    }
    const [salesData, seasonsData, unitsData] = await Promise.all([
      salesRes.json(),
      seasonsRes.json(),
      unitsRes.json(),
    ]);
    const data = salesData.sale.map((sale, index) => {
      const {
        cantidad_vendida: cantidadVendida,
        fecha_venta: fechaVenta,
        id,
        id_temporada: idTemporada,
        id_unidad_medida: idUnidadMedida,
        observaciones,
        precio_total: precioTotal,
        precio_unitario: precioUnitario,
      } = sale;
      const season = seasonsData.find((season) => season.id === idTemporada);
      const unit = unitsData.units.find((unit) => unit.id === idUnidadMedida);

      return {
        id: index + 1,
        temporada: season.nombre,
        fechaVenta,
        cantidadVendida,
        unidadMedida: unit.nombre,
        precioUnitario,
        precioTotal,
        observaciones,
        idSale: id,
        idTemporada,
      };
    });
    return data;
  } catch (error) {
    console.error("error fetching data", error);
    return (data = []);
  }
}
