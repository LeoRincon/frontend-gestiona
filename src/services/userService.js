import { API_URL, USERS_PATH, LOGIN_PATH } from "../utils/const";

const url = API_URL + USERS_PATH;
const loginUrl = API_URL + LOGIN_PATH

export async function fetchUsers() {
  const response = await fetch(url, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
  });

  const { data } = await response.json();

  const newdata = data.map((user) => {
    return {
      name: user.nombre,
      email: user.email,
    };
  });
  return newdata;
}

export async function createUser(data) {
  if (!data || !data.nombre || !data.email || !data.password) {
    throw new Error("Incomplete data");
  }

  const response = await fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  const result = await response.json();

  return result;
}

export async function loginUser(data){
  if (!data || !data.email || !data.password) {
    throw new Error("Incomplete data");
  }
  
  const response = await fetch(loginUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json"
    },
    body: JSON.stringify(data)
  })

  const result = await response.json();
  if(result.error) throw new Error(result.error)
  
  return result
}
