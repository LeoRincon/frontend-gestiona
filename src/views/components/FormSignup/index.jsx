import { useState } from "react";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router";
import { formSignupTexts } from "../../../utils/const";
import { createUser } from "../../../services/userService";

import "./styles.css";

const { fullname, email, password, createCount } = formSignupTexts;

export default function FormSignup() {
  const { register, handleSubmit, reset } = useForm();
  const navigate = useNavigate();
  const [errorMessage, setErrorMessage] = useState(""); // Estado para manejar errores

  const onSubmit = async (data) => {
    try {
      const result = await createUser(data);
      if (!result.success) {
        throw new Error(result.message);
      }
      reset();
      navigate("/");
    } catch (error) {
      setErrorMessage(error.message || "Error al crear la cuenta");
    }
  };

  return (
    <form className="cardform__form" onSubmit={handleSubmit(onSubmit)}>
      <label className="form__label--required" htmlFor="nombre">
        {fullname}
      </label>
      <input
        className="form__input"
        type="text"
        id="fullname"
        name="nombre"
        placeholder="Ingrese su nombre completo"
        {...register("nombre", { required: true })}
        required
      />
      <label className="form__label--required" htmlFor="email">
        {email}
      </label>
      <input
        className="form__input"
        type="email"
        id="email"
        name="email"
        placeholder="Ingrese su correo"
        {...register("email", { required: true })}
        required
      />
      <label className="form__label--required" htmlFor="password">
        {password}
      </label>
      <input
        className="form__input"
        type="password"
        id="password"
        name="password"
        placeholder="Ingrese su contraseÃ±a"
        {...register("password", { required: true })}
        required
      />
      <button className="form__button" type="submit">
        {createCount}
      </button>
      
      {/* Mensaje de error */}
      {errorMessage && <p className="error-message">{errorMessage}</p>}
    </form>
  );
}
