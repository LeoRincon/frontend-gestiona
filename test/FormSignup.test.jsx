import { test, describe, beforeEach, vi, expect } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FormSignup from '../src/views/components/FormSignup';
import { WrapperTest } from './WrapperTest';
import * as userService from '../src/services/userService';

vi.mock('../src/services/userService', () => ({
 createUser: vi.fn(),
}));

describe('FormSignup Component', () => {
 beforeEach(() => {
  vi.clearAllMocks();
 });

 test('Debe permitir el registro exitoso', async () => {
  userService.createUser.mockResolvedValue({ success: true });

  render(
   <WrapperTest>
    <FormSignup />
   </WrapperTest>
  );

  const nameInput = screen.getByPlaceholderText('Ingrese su nombre completo');
  const emailInput = screen.getByPlaceholderText('Ingrese su correo');
  const passwordInput = screen.getByPlaceholderText('Ingrese su contraseña');
  const signupButton = screen.getByRole('button', { name: /crear cuenta/i });

  await userEvent.type(nameInput, 'Juan Pérez');
  await userEvent.type(emailInput, 'juan@example.com');
  await userEvent.type(passwordInput, 'password123');

  await userEvent.click(signupButton);

  await waitFor(() => {
   expect(userService.createUser).toHaveBeenCalledWith({
    nombre: 'Juan Pérez',
    email: 'juan@example.com',
    password: 'password123',
   });
  });
 });

 test('Debe mostrar un error si el registro falla', async () => {
  userService.createUser.mockResolvedValue({
   success: false,
   message: 'Error en el registro',
  });

  render(
   <WrapperTest>
    <FormSignup />
   </WrapperTest>
  );

  const nameInput = screen.getByPlaceholderText('Ingrese su nombre completo');
  const emailInput = screen.getByPlaceholderText('Ingrese su correo');
  const passwordInput = screen.getByPlaceholderText('Ingrese su contraseña');
  const signupButton = screen.getByRole('button', { name: /crear cuenta/i });

  await userEvent.type(nameInput, 'Juan Pérez');
  await userEvent.type(emailInput, 'juan@example.com');
  await userEvent.type(passwordInput, 'password123');

  await userEvent.click(signupButton);

  expect(await screen.findByText(/error en el registro/i)).toBeInTheDocument();
 });
});
