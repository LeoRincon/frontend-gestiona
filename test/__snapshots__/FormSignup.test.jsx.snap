// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormSignup Component > Debe mostrar un error si el registro falla 1`] = `
<div>
  <form
    class="cardform__form"
  >
    <label
      class="form__label--required"
      for="nombre"
    >
      Nombre Completo
    </label>
    <input
      class="form__input"
      id="fullname"
      name="nombre"
      placeholder="Ingrese su nombre completo"
      required=""
      type="text"
    />
    <label
      class="form__label--required"
      for="email"
    >
      Correo
    </label>
    <input
      class="form__input"
      id="email"
      name="email"
      placeholder="Ingrese su correo"
      required=""
      type="email"
    />
    <label
      class="form__label--required"
      for="password"
    >
      Contrase単a
    </label>
    <input
      class="form__input"
      id="password"
      name="password"
      placeholder="Ingrese su contrase単a"
      required=""
      type="password"
    />
    <button
      class="form__button"
      type="submit"
    >
      Crear Cuenta
    </button>
    <p
      class="error-message"
    >
      Error en el registro
    </p>
  </form>
</div>
`;

exports[`FormSignup Component > Debe permitir el registro exitoso 1`] = `
<div>
  <form
    class="cardform__form"
  >
    <label
      class="form__label--required"
      for="nombre"
    >
      Nombre Completo
    </label>
    <input
      class="form__input"
      id="fullname"
      name="nombre"
      placeholder="Ingrese su nombre completo"
      required=""
      type="text"
    />
    <label
      class="form__label--required"
      for="email"
    >
      Correo
    </label>
    <input
      class="form__input"
      id="email"
      name="email"
      placeholder="Ingrese su correo"
      required=""
      type="email"
    />
    <label
      class="form__label--required"
      for="password"
    >
      Contrase単a
    </label>
    <input
      class="form__input"
      id="password"
      name="password"
      placeholder="Ingrese su contrase単a"
      required=""
      type="password"
    />
    <button
      class="form__button"
      type="submit"
    >
      Crear Cuenta
    </button>
  </form>
</div>
`;
